src/node_crypto.cc

```diff
@@ -226,6 +226,43 @@ static int CryptoPemCallback(char *buf, int size, int rwflag, void *u) {
   return 0;
 }
 
+// Loads OpenSSL engine by engine id and returns it. The loaded engine
+// gets a reference so remember the corresponding call to ENGINE_free
+// In case of error the appropriate js exception is schduled
+// and nullptr is returned.
+#ifndef OPENSSL_NO_ENGINE
+static ENGINE* LoadEngineById(Environment* env, const char *engine_id) {
+  ENGINE* engine = ENGINE_by_id(engine_id);
+
+  if (engine == nullptr) {
+    // Engine not found, try loading dynamically
+    engine = ENGINE_by_id("dynamic");
+    if (engine != nullptr) {
+      if (!ENGINE_ctrl_cmd_string(engine, "SO_PATH", engine_id, 0) ||
+          !ENGINE_ctrl_cmd_string(engine, "LOAD", nullptr, 0)) {
+        ENGINE_free(engine);
+        engine = nullptr;
+      }
+    }
+  }
+
+  if (engine == nullptr) {
+    char errmsg[1024];
+    int err = ERR_get_error();
+    if (err != 0) {
+        ERR_error_string_n(err, errmsg, sizeof(errmsg));
+    } else {
+      snprintf(errmsg, sizeof(errmsg),
+               "Engine \"%s\" was not found", engine_id);
+    }
+    env->ThrowError(errmsg);
+    // Forcibly clear OpenSSL's error stack
+    ERR_clear_error();
```

Comment:
Yes. I was not aware of that struct. The reason I did ERR_clear_error manually was due to the comment for struct ClearErrorOnReturn saying "ERR_clear_error() isn't necessarily cheap" so I thought that in the 'golden path' case I would prefer not to call it (which the struct's will always do). I guess that in the big picture of loading a .so file from disk etc. calling ERR_clear_error or ERR_pop_to_mark is insignificant - it's just an old habit not to do stuff that is not required...